### PR TITLE
Enable suggestion mode for LSP completion

### PR DESCRIPTION
**Pull request by dibarbet**
Resolves https://github.com/dotnet/aspnetcore/issues/22754

### CODE REVIEW

**Comment by dibarbet** [comment]
cc @NTaylorMullen

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs [hunk 1]
```diff
@@ -90,12 +90,16 @@ public override async Task<LSP.CompletionItem[]> HandleRequestAsync(LSP.Completi
...
+                return null;
             }
 
             var lspVSClientCapability = context.ClientCapabilities?.HasVisualStudioLspCapability() == true;
 
-            return list.Items.Select(item => CreateLSPCompletionItem(request, item, lspVSClientCapability, completionTrigger)).ToArray();
+            return new LSP.VSCompletionList
+            {
+                Items = list.Items.Select(item => CreateLSPCompletionItem(request, item, lspVSClientCapability, completionTrigger)).ToArray(),
+                SuggesstionMode = list.SuggestionModeItem != null,
```
Comment (allisonchou) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:101]: Not super important, would it be good to file a bug on LSP for the misspelling of this word?

**Review by davidwengier — APPROVED** [review]

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs (hunk 1 above)
Comment (NTaylorMullen) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:101]: Out of curiosity does `SuggestionModeItem` only ever equal not equal `null` at contextually relevant locations? Just trying to make sure it doesn't impact the Razor scenario where you type `@Dat|` and can complete `DateTime` by typing `.`.

Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:101]: Based on the comment, it's only non-null when the list should be operating in suggestion mode - http://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/Completion/CompletionContext.cs,136

Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:101]: for misspelling https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1192571

File: src/Features/Core/Portable/Completion/Providers/AbstractEmbeddedLanguageCompletionProvider.cs [hunk 2]
```diff
@@ -32,11 +34,11 @@ protected AbstractEmbeddedLanguageCompletionProvider()
         {
         }
 
-        protected ImmutableArray<IEmbeddedLanguage> GetLanguageProviders<T>(Func<T, Document> documentProvider, T state)
+        protected ImmutableArray<IEmbeddedLanguage> GetLanguageProviders<T>(Func<T?, Document?> documentProvider, T state)
```
Comment (sharwell) [inline src/Features/Core/Portable/Completion/Providers/AbstractEmbeddedLanguageCompletionProvider.cs:37]:
```suggestion
        protected ImmutableArray<IEmbeddedLanguage> GetLanguageProviders<T>(Func<T, Document?> documentProvider, T state)
```

File: src/Features/Core/Portable/Completion/Providers/AbstractEmbeddedLanguageCompletionProvider.cs [hunk 3]
```diff
@@ -63,7 +65,7 @@ public override bool ShouldTriggerCompletion(SourceText text, int caretPosition,
 
         public override async Task ProvideCompletionsAsync(CompletionContext context)
         {
-            foreach (var language in GetLanguageProviders(context => context.Document, context))
+            foreach (var language in GetLanguageProviders(context => context?.Document, context))
```
Comment (sharwell) [inline src/Features/Core/Portable/Completion/Providers/AbstractEmbeddedLanguageCompletionProvider.cs:68]:
```suggestion
            foreach (var language in GetLanguageProviders(context => context.Document, context))
```

File: src/Features/Core/Portable/Completion/Providers/AbstractEmbeddedLanguageCompletionProvider.cs [hunk 4]
```diff
@@ -45,7 +47,7 @@ protected ImmutableArray<IEmbeddedLanguage> GetLanguageProviders<T>(Func<T, Docu
 
         public override bool ShouldTriggerCompletion(SourceText text, int caretPosition, CompletionTrigger trigger, OptionSet options)
         {
-            foreach (var language in GetLanguageProviders(text => text.GetOpenDocumentInCurrentContextWithChanges(), text))
+            foreach (var language in GetLanguageProviders(text => text?.GetOpenDocumentInCurrentContextWithChanges(), text))
```
Comment (sharwell) [inline src/Features/Core/Portable/Completion/Providers/AbstractEmbeddedLanguageCompletionProvider.cs:50]:
```suggestion
            foreach (var language in GetLanguageProviders(text => text.GetOpenDocumentInCurrentContextWithChanges(), text))
```

**Pull request merged**